libcudacxx/include/cuda/__bit/ffs.h

```diff
@@ -0,0 +1,175 @@
+#include <cuda/std/limits>
+
+#if _CCCL_COMPILER(MSVC)
+#  include <intrin.h>
+#endif // _CCCL_COMPILER(MSVC)
+
+#include <cuda/std/__cccl/prologue.h>
+
+_CCCL_BEGIN_NAMESPACE_CUDA
+
+#if _CCCL_HAS_BUILTIN(__builtin_ffs) || _CCCL_COMPILER(GCC)
+#  define _CCCL_BUILTIN_FFS(...)   __builtin_ffs(__VA_ARGS__)
+#  define _CCCL_BUILTIN_FFSLL(...) __builtin_ffsll(__VA_ARGS__)
+#endif // _CCCL_HAS_BUILTIN(__builtin_ffs) || _CCCL_COMPILER(GCC)
+
+template <typename _Tp>
+[[nodiscard]] _CCCL_API constexpr int __ffs_impl_constexpr(_Tp __v) noexcept
+{
+  static_assert(::cuda::std::__cccl_is_unsigned_integer_v<_Tp>, "_Tp must be unsigned");
+
+  if (__v == 0)
+  {
+    return 0;
+  }
+
+  int __pos = 1;
+  while ((__v & 1) == 0)
+  {
+    __v >>= 1;
+    ++__pos;
+  }
+  return __pos;
+}
+
+#if !_CCCL_COMPILER(NVRTC)
+template <typename _Tp>
+[[nodiscard]] _CCCL_API _CCCL_HOST int __ffs_impl_host(_Tp __v) noexcept
+{
+#  if defined(_CCCL_BUILTIN_FFS)
+  if constexpr (sizeof(_Tp) <= sizeof(int))
+  {
+    return _CCCL_BUILTIN_FFS(static_cast<int>(__v));
+  }
+  else
+  {
+    return _CCCL_BUILTIN_FFSLL(static_cast<long long>(__v));
+  }
+#  elif _CCCL_COMPILER(MSVC)
+  unsigned long __where{};
+  unsigned char __res{};
+  if constexpr (sizeof(_Tp) <= sizeof(uint32_t))
+  {
+    __res = ::_BitScanForward(&__where, static_cast<uint32_t>(__v));
+  }
+  else
+  {
+    __res = ::_BitScanForward64(&__where, static_cast<uint64_t>(__v));
+  }
+  return static_cast<int>(__where) + (__res != 0);
+#  else
+  return ::cuda::__ffs_impl_constexpr(__v);
+#  endif // _CCCL_COMPILER(MSVC)
+}
+#endif // !_CCCL_COMPILER(NVRTC)
+
+#if _CCCL_CUDA_COMPILATION()
+template <typename _Tp>
+[[nodiscard]] _CCCL_DEVICE_API int __ffs_impl_device(_Tp __v) noexcept
+{
+  if constexpr (sizeof(_Tp) <= sizeof(int))
+  {
+    return ::__ffs(static_cast<int>(__v));
+  }
+  else
+  {
+    return ::__ffsll(static_cast<long long>(__v));
+  }
+}
+#endif // _CCCL_CUDA_COMPILATION()
+
+_CCCL_TEMPLATE(typename _Tp)
+_CCCL_REQUIRES(::cuda::std::__cccl_is_unsigned_integer_v<_Tp>)
+[[nodiscard]] _CCCL_API constexpr int ffs(_Tp __v) noexcept
+{
+#if _CCCL_HAS_INT128()
+  if constexpr (sizeof(_Tp) == sizeof(__uint128_t))
+  {
+    const auto __lo = static_cast<uint64_t>(__v);
+    const auto __hi = static_cast<uint64_t>(static_cast<__uint128_t>(__v) >> 64);
+
+    if (!::cuda::std::__cccl_default_is_constant_evaluated())
+    {
+      if (__lo != 0)
+      {
+        NV_IF_ELSE_TARGET(
+          NV_IS_HOST, (return ::cuda::__ffs_impl_host(__lo);), (return ::cuda::__ffs_impl_device(__lo);));
+      }
+      if (__hi != 0)
+      {
+        NV_IF_ELSE_TARGET(
+          NV_IS_HOST, (return ::cuda::__ffs_impl_host(__hi) + 64;), (return ::cuda::__ffs_impl_device(__hi) + 64;));
+      }
+      return 0;
+    }
+
+    if (__lo != 0)
+    {
+      return ::cuda::__ffs_impl_constexpr(__lo);
+    }
+    if (__hi != 0)
+    {
+      return ::cuda::__ffs_impl_constexpr(__hi) + 64;
+    }
+    return 0;
```

Comment:
```suggestion
    const auto __lo = static_cast<::cuda::std::uint64_t>(__v);
    const auto __hi = static_cast<::cuda::std::uint64_t>(static_cast<__uint128_t>(__v) >> 64);
    
    if (const auto __result = ::cuda::ffs(__lo) != 0)
    {
      return __result;
    }
    if (const auto __result = ::cuda::ffs(__hi) != 0)
    {
      return __result + 64;
    }
    return 0;
```
can be simplified to